### PR TITLE
Replace the trustedcaller hack with a debugger command

### PR DESCRIPTION
**Pull request by Warepire**
This makes the check safer as it doesn't rely on stack frames being within specific sizes.

Fixes issue #52

### CODE REVIEW

File: source/application/wintaser.cpp [hunk 1]
```diff
@@ -2885,177 +2874,79 @@ LPCWSTR ExceptionCodeToDescription(DWORD code)
...
-		{
-			SetTrustedRangeInfo(infos.infos[1+i], trustedModuleInfos[i]);
-			infos.numInfos++;
-		}
-	}
-	
-	SIZE_T bytesWritten = 0;
-	WriteProcessMemory(hProcess, remoteTrustedRangeInfos, &infos, sizeof(infos), &bytesWritten);
-}
+static std::map<DWORD, std::vector<std::wstring>> trusted_modules;
```
Comment (YaLTeR) [inline source/application/wintaser.cpp]: Should be `gs_trusted_modules`.

File: source/application/wintaser.cpp [hunk 2]
```diff
@@ -2885,177 +2874,79 @@ LPCWSTR ExceptionCodeToDescription(DWORD code)
...
-	int count = trustedModuleInfos.size();
-	for(int i = 0; i < count; i++)
-		if(IsInRange(address, trustedModuleInfos[i].mi))
-			return true;
-	return (address == 0);
+    return std::count(str.begin(), str.end(), L'\\');
 }
 
-DWORD CalculateModuleSize(LPVOID hModule, HANDLE hProcess)
+static void ModuleLoaded(DWORD process_id, const std::wstring& path)
```
Comment (YaLTeR) [inline source/application/wintaser.cpp]: Perhaps `OnModuleLoaded`? The current name can be confused with a function that checks if the given module is loaded.

File: source/application/wintaser.cpp [hunk 3]
```diff
@@ -2885,177 +2874,79 @@ LPCWSTR ExceptionCodeToDescription(DWORD code)
...
-		{
-			if(trustedModuleInfos[i].mi.lpBaseOfDll == hModule)
-			{
-				trustedModuleInfos.erase(trustedModuleInfos.begin() + i);
-				SendTrustedRangeInfos(hProcess);
-				break;
-			}
-		}
-	}
+static bool IsTrustedModule(DWORD process_id, const std::wstring& path)
```
Comment (YaLTeR) [inline source/application/wintaser.cpp]: Maybe all this stuff needs to move from `wintaser.cpp` into some `trusted_module.cpp`? I'm not a huge fan of the huge `wintase{e,r}.cpp` files with a lot of functions that could well be grouped into separate files.

File: source/application/wintaser.cpp [hunk 4]
```diff
@@ -2885,177 +2874,79 @@ LPCWSTR ExceptionCodeToDescription(DWORD code)
...
+        if (path == m)
+        {
+            return true;
+        }
+        return false;
+    };
+    auto result = std::find_if(trusted_modules[process_id].begin(),
+                               trusted_modules[process_id].end(),
+                               CheckModule);
+    return (result != trusted_modules[process_id].end());
```
Comment (YaLTeR) [inline source/application/wintaser.cpp]: All this function => `std::find(<...>, path) != <...>.cend()`

File: source/application/wintaser.cpp [hunk 5]
```diff
@@ -3942,6 +3826,41 @@ static DWORD WINAPI DebuggerThreadFunc(LPVOID lpParam)
                                             }
                                         }
                                         break;
+                                    case IPC::Command::CMD_IS_TRUSTED_CALLER:
+                                        {
+                                            // print the callstack of the thread
```
Comment (YaLTeR) [inline source/application/wintaser.cpp]: I think this doesn't belong here any more.

**Review by YaLTeR — CHANGES_REQUESTED** [review]

File: source/application/wintaser.cpp [hunk 6]
```diff
@@ -2885,177 +2874,79 @@ LPCWSTR ExceptionCodeToDescription(DWORD code)
...
-			{
-				debugprintf(L"apparently already TRUSTED MODULE 0x%08X - 0x%08X (%s)\n", mmi.mi.lpBaseOfDll, (DWORD)mmi.mi.lpBaseOfDll+mmi.mi.SizeOfImage, path);
-				return;
-			}
-		}
-		trustedModuleInfos.push_back(mmi);
-		SendTrustedRangeInfos(hProcess);
-	}
-	debugprintf(L"TRUSTED MODULE 0x%08X - 0x%08X (%s)\n", mmi.mi.lpBaseOfDll, (DWORD)mmi.mi.lpBaseOfDll+mmi.mi.SizeOfImage, path);
+    if (removed == proc_modules.end())
```
Comment (YaLTeR) [inline source/application/wintaser.cpp]: I'm pretty sure this check isn't needed, `.erase()` will work correctly regardless.

Comment (Warepire) [reply source/application/wintaser.cpp]: I got interator-out-of-range asserts.

Comment (YaLTeR) [reply source/application/wintaser.cpp]: Oh okay

File: source/application/TrustedModule.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace
+{
+    std::map<DWORD, std::vector<std::wstring>> s_trusted_modules;
+
+    UINT CountSharedPrefixLength(const std::wstring& a, const std::wstring& b)
+    {
+        size_t shortest_path = std::min(a.size(), b.size());
+        UINT i;
+        for (i = 0; i < shortest_path && tolower(a[i]) == tolower(b[i]); ++i);
```
Comment (YaLTeR) [inline source/application/TrustedModule.cpp]: I think this should be
```
for (size_t i = 0; /* ... */)
    ;
```
- move the `i` type into the loop and change it to `size_t`
- put `;` on a separate line so it stands out

File: source/application/TrustedModule.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        return i;
+    }
+    UINT CountBackslashes(const std::wstring& str)
```
Comment (YaLTeR) [inline source/application/TrustedModule.cpp]: A blank line between functions would be nice.

File: source/application/TrustedModule.cpp [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+extern std::wstring injected_dll_path;
+extern std::wstring sub_exe_filename;
+
+namespace TrustedModule
+{
+    void OnLoad(DWORD process_id, const std::wstring& path)
+    {
+        /*
+        * If we can "trust" a newly loaded module, add it to the list.
+        */
```
Comment (YaLTeR) [inline source/application/TrustedModule.cpp]: These two lines should be one space to the right.

File: source/application/TrustedModule.cpp [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        UINT proximity = 2;
+        if (!wcsicmp(path.c_str(), injected_dll_path.c_str()))
```
Comment (YaLTeR) [inline source/application/TrustedModule.cpp]: Change this to `path != injected_dll_path`.

File: source/application/TrustedModule.cpp [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        }
+        else if (CountBackslashes(Config::exe_filename.c_str() + CountSharedPrefixLength(Config::exe_filename, path)) < proximity)
+        {
+            s_trusted_modules[process_id].emplace_back(path);
+        }
+        else if (CountBackslashes(sub_exe_filename.c_str() + CountSharedPrefixLength(sub_exe_filename, path)) < proximity)
+        {
+            s_trusted_modules[process_id].emplace_back(path);
+        }
+        else if (path.length() >= 4 && !wcsnicmp(path.c_str() + path.length() - 4, L".cox", 4)) // hack, we can generally assume the game outputted any dll that has this extension
```
Comment (YaLTeR) [inline source/application/TrustedModule.cpp]: Perhaps `path.length() >= 4 && path.substr(path.length() - 4) == ".cox"`

File: source/application/TrustedModule.cpp [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        {
+            s_trusted_modules[process_id].emplace_back(path);
+        }
+    }
+
+    void OnUnload(DWORD process_id, const std::wstring& path)
+    {
+        auto& proc_modules = s_trusted_modules[process_id];
+        auto removed = std::remove_if(proc_modules.begin(), proc_modules.end(),
+            [&path](const std::wstring& m) { return (m.compare(path) == 0); });
```
Comment (YaLTeR) [inline source/application/TrustedModule.cpp]: `m == path`

File: source/application/TrustedModule.cpp [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    bool IsTrusted(DWORD process_id, const std::wstring& path)
+    {
+        auto CheckModule = [&path](const std::wstring &m) {
+            return (path == m);
+        };
+        return (std::find_if(s_trusted_modules[process_id].begin(),
+                s_trusted_modules[process_id].end(),
+                CheckModule) != s_trusted_modules[process_id].end());
```
Comment (YaLTeR) [inline source/application/TrustedModule.cpp]: I think these two lines should be aligned like this:
```
return (std::find_if(s_trusted_modules[process_id].begin(),
                     s_trusted_modules[process_id].end(),
                     CheckModule) != s_trusted_modules[process_id].end());
```

File: source/application/TrustedModule.cpp (hunk 7 above)
Comment (Warepire) [reply source/application/TrustedModule.cpp]: I'll add an empty `{}` so it's clearer. I have to return the `i`, so it ends up outside the loop. Can definitely make it `size_t`.

File: source/application/TrustedModule.cpp (hunk 8 above)
Comment (Warepire) [reply source/application/TrustedModule.cpp]: I agree.

File: source/application/TrustedModule.cpp (hunk 10 above)
Comment (Warepire) [reply source/application/TrustedModule.cpp]: It's `wcsicmp`, case insensitive. I don't see why it needs to be that now though. I will change it.

File: source/application/TrustedModule.cpp (hunk 11 above)
Comment (Warepire) [reply source/application/TrustedModule.cpp]: Yes.

File: source/application/TrustedModule.cpp (hunk 12 above)
Comment (Warepire) [reply source/application/TrustedModule.cpp]: I coded way too much C today... I will fix that.

File: source/application/TrustedModule.cpp (hunk 13 above)
Comment (Warepire) [reply source/application/TrustedModule.cpp]: They were at one point... Will fix it.

File: source/application/TrustedModule.cpp (hunk 7 above)
Comment (YaLTeR) [reply source/application/TrustedModule.cpp]: Right, good point about returning. Somehow I didn't think about that.

**Review by YaLTeR — APPROVED** [review]

**Pull request merged**